Fix/decl const cpp api

bgd54 opened the pull request:
Fix issue with OPS_instance::decl_const:
This function called a function that does not exist. Changed to call ops_decl_const_char and perform the same type checking as ops_decl_const.

Also cleans up unnecessary ops_decl_const2 function from code generator.

ops_instance.h:
  remove call to nonexisting function _ops_decl_const
  call ops_decl_const_char instead
ops_internal1.h:
  Add OPS_instance parameter to ops_decl_const_char
ops_internal2.h:
  Remove function declaration: function already declared in
  ops_internal1.h
  Remove ops_decl_const2: does the same as ops_decl_const
ops_lib_core.h:
  add OPS_instance parameters to ops_decl_const_char calls
ops_mpi_decl.cpp, ops_seq.cpp:
  Add check on instance parameter, call ops_execute on the correct
  instance.

ops.py:
  remove code replace on ops_decl_const (ops_decl_const2 removed)

ops_gen_mpi_inline.py:
  Remove ops_decl_const_char2 - never used.
ops_gen_mpi_<target>.py:
  Add instance parameter to generated ops_decl_const_char funcitons

reguly commented:
All that handling of instance being nullptr doesn't look so nice. I think we can safely just pass OPSInstance::getOPSInstance() to ops_decl_const_char if the C API ops_decl_const or ops_update_const is called. That would cut down on the subsequent checks.

bgd54 commented:
Yeah we haven't included ops_instance.h where we define ops_decl_const yet. So we don't have access to the getter (and the instance depends on the content of ops_lib_core.h). I've added a helper function in place of the original ops_decl_const2 to call ..._char with the instance and removed the nullptr checks from ops_decl_const_char functions.

reguly reviewed (approved):
Looks good to me. @gihanmudalige can you run the full suite of tests?

reguly commented:
Something is still wrong with ops_decl_const parsing, the CI seems to fail.

bgd54 commented:
Sorry, forgot to change the group for the new regex. I think everything should work now.

reguly commented:
@gihanmudalige Can you run the full test suite on this please?

gihanmudalige commented:
@reguly yes, will do now.
